wayland: Add support for images in clipboard

LiquidityC opened the pull request:
Re-writes the clipboard data handling in wayland to an on demand solution where callbacks are provided to generate/provide the clipboard data when requested by the OS.

Introduces:
- SDL_GetClipboardUserdata
- SDL_GetClipboardData
- SDL_SetClipboardData
- SDL_HasClipboardData

## Description

As well as implementing the added functions in the wayland driver the handling of clipboard content in wayland has been changed to instead hold callbacks that provide the data on demand when the OS requests it.
This behavior is in line with the new clipboard api functions.

## Existing Issue(s)

- #7625

LiquidityC commented on include/SDL3/SDL_clipboard.h:
```diff
@@ -129,6 +129,89 @@ extern DECLSPEC char * SDLCALL SDL_GetPrimarySelectionText(void);
+ * - Before calling SDL_Quit()
+ *
+ * \param callback      A function pointer to the function that provides the clipboard data
+ * \param mime_types    A list of mime-types that are being offered. Must end with a NULL value
```
How are we feeling about the "String list must end with a NULL guard" situation?
I could just switch to a `mime_count` argument.

slouken replied:
Yeah, let's use a count argument for this.

LiquidityC commented on include/SDL3/SDL_clipboard.h:
```diff
@@ -129,6 +129,89 @@ extern DECLSPEC char * SDLCALL SDL_GetPrimarySelectionText(void);
 extern DECLSPEC SDL_bool SDLCALL SDL_HasPrimarySelectionText(void);
 
+/**
+ * Callback function that will be called when data for the specified mime-type
+ * is requested by the OS.
+ *
+ * \param size      The length of the returned data
+ * \param mime_type The requested mime-type
+ * \param userdata  A pointer to provided user data
+ * \returns a pointer to the data for the provided mime-type
```
I'm thinking NULL response. Its not really a valid implementation but I'll see how Wayland handles it.

LiquidityC replied:
Seems like the receiving application needs to deal with it. Discord handled NULL data for an image gracefully. Gimp sort of soft-locked. I guess it's sort of UB...? I'll see if I can find some docs.

LiquidityC replied:
@Guldoman Spotted that we omit closing the FD when data is NULL or length is zero. After fixing that. Returning NULL or returning length zero didn't cause flaky behavior in receiving apps. It's still up to the receiving app to deal with such a case but it does seem to work "nicer". I'll make a note of it in doc.

LiquidityC replied:
The actual answer to your question is probably: If you don't handle a certain mime-type you shouldn't announce that you do. Eg. We shouldn't be getting requests for mime-types that we didn't first announce that we had available. So returning NULL or 0 is really poor form.

Guldoman commented on src/video/wayland/SDL_waylanddatamanager.c:
```diff
@@ -136,29 +137,20 @@ static ssize_t read_pipe(int fd, void **buffer, size_t *total_length, SDL_bool n
+            // Just keep spinning
         }
+        close(fd);
+        SDL_free(data);
```
Is this freeing data received from the callback?

LiquidityC replied:
Yes

Guldoman replied:
I would avoid doing that and just let the application deal with freeing it when it needs to. Otherwise the application would have to create a copy of the data every time a paste operation is needed.

LiquidityC replied:
Yes... it does put a bit more demand on the application implementation though. But I also think it's a more sound approach. Just awkward having to retain and free returned data. I'll fix

Guldoman commented on src/video/wayland/SDL_waylanddatamanager.c:
```diff
@@ -136,29 +137,20 @@ static ssize_t read_pipe(int fd, void **buffer, size_t *total_length, SDL_bool n
+        while (write_pipe(fd, data, length, &result) > 0) {
+            // Just keep spinning
         }
+        close(fd);
```
Could it be a good idea to close the pipe even if `data` is `NULL` or `length` is 0?

LiquidityC replied:
Definitely. That's probably why it's causing receiving apps to be flaky. Good spot.

LiquidityC commented:
The failed checks seem to be connection issues to Azure. I can't restart the failed ones but perhaps someone can?

Guldoman commented on src/video/wayland/SDL_waylandclipboard.c:
```diff
@@ -22,10 +22,66 @@
+    "STRING",
+};
+
+static void *Wayland_ClipboardTextCallback(size_t *length, const char *mime_type, void *userdata)
+{
+    void *data = NULL;
+    *length = 0;
+
+    if (userdata != NULL) {
+        char *text = userdata;
+        *length = SDL_strlen(text);
+        data = userdata;
+    }
+
+    return data;
+}
```
I wonder if it would be a good idea to check the mime-type anyways. Can applications request unadvertised mime-types?

LiquidityC replied:
I can't find any definite answer in the docs if Wayland blocks this or we have to deal with it so we might as well check on our side to be safe.

LiquidityC commented on include/SDL3/SDL_clipboard.h:
```diff
@@ -129,6 +129,96 @@ extern DECLSPEC char * SDLCALL SDL_GetPrimarySelectionText(void);
+ *
+ * \param callback      A function pointer to the function that provides the clipboard data
+ * \param mime_count    The number of mime-types in the mime_types list
+ * \param mime_types    A list of mime-types that are being offered. Must end with a NULL value
```
Nope. Forgot to update the doc.

LiquidityC commented:
Can this one be merged or is it preferred that I add X11 to this PR as well?

LiquidityC commented:
Rebased and squashed. I'll get cracking on X11.

slouken commented on include/SDL3/SDL_events.h:
```diff
@@ -530,6 +531,18 @@ typedef struct SDL_DropEvent
+    Uint32 type;        /**< ::SDL_EVENT_CLIPBOARD_CANCELLED or ::SDL_EVENT_CLIPBOARD_UPDATE */
+    Uint64 timestamp;   /**< In nanoseconds, populated using SDL_GetTicksNS() */
+    void *userdata;     /**< User data if any has been set. NULL for ::SDL_EVENT_CLIPBOARD_UPDATE */
+} SDL_ClipboardCancelled;
```
Suggest this be called SDL_ClipboardEvent, since it's used by the update and canceled events.

LiquidityC replied:
Agreed

slouken commented on src/video/SDL_clipboard.c:
```diff
@@ -22,6 +22,23 @@
+
+    if (_this->SetClipboardData) {
+        status = _this->SetClipboardData(_this, callback, mime_count, mime_types, userdata);
+    }
```
Suggest returning SDL_Unsupported() if _this->SetClipboardData isn't provided?
Alternatively, the clipboard data can be managed internally so at least the application can provide copy/paste of multiple mime-types to itself.

LiquidityC replied:
I'll go with unsupported in this PR. I'll take a look at potentially handling it internally as I'm implementing X11.

Kontrabant reviewed (approved):
Aside from some minor nitpicks, the Wayland portion of the code looks good to me.

LiquidityC commented on src/video/wayland/SDL_waylandclipboard.c:
```diff
@@ -22,10 +22,78 @@
+    int status = 0;
+
+    if (_this == NULL || _this->driverdata == NULL) {
+        status = SDL_SetError("Video driver uninitialized");
```
I assumed as much but this was the pattern adopted in the existing code so I chose to follow suit.

It became abundantly clear that they were redundant now that I dove into the X11 driver. 

I'll address this and your other comments tonight.

LiquidityC commented:
I addressed the review comments. However GitHub acations seems a bit sad today. If someone with permissions to do so could re-start the jobs at a later stage I'm quite confident that they should go through (famous last words).

LiquidityC commented:
I have a working draft of X11 ready. Personally I'd prefer if this PR could be merged and I could create a new one based on main. But if you'd rather have X11 on top of this one I can oblige.

X11 hasn't required any changes to the new clipboard api introduced in this PR.

LiquidityC commented:
Nice. Thanks.

I sumitted a PR for X11. Assume I'll have some work before the tests start going through.

I'll start looking into an internal handling when driver support is lacking. Then I will try and get my hands on a Windows machine (might be able to borrow one from work). As for Mac... that's going to be tricky... Maybe I can find a colleague with some time on his/her hands... I'll look around.
